### DISCUSSION TITLE
How to pass form data to queryClient.setMutationDefault when submitting a form

### DISCUSSION THREAD

**Discussion by aubreyzulu** [TanStack/query#1598] [Q&A]
How do i pass form data to queryClient.setMutationDefault when submitting a form..?

**Comment by TkDodo** [comment]
data is always passed to the mutation function (mutate or mutateAsync) returned by useMutation, not to useMutation itself. So, i don’t think i understand the question 🙃

**Reply by aubreyzulu** [reply]
Here is sample code `
queryClient.setMutationDefaults('user/register', {
  mutationFn: async (data) => {
    console.log(data);
    try {
      const response = await axios.post('/user/register', data);
      return response.data;
    } catch (error) {
      if (error.response) {
        throw new HttpError(
          error.message,
          error.response.status,
          error.response.data
        );
      }
      throw new HttpError('something went wrong');
    }
  }
});

form code 
const { mutate, isLoading } = useMutation(
    (data) => axios.post('/user/register', data),
  );

const onSubmit = (data) =>{mutate(data);};

`

**Reply by TkDodo** [reply]
If you set a mutationFn as default, why do you also pass a function to useMutation? That overrides the default.

**Comment by aubreyzulu** [comment]
I removed it, it's still not working. only working when I give a function
to useMutation, but I want to have the function passed to query liebt.
setMutationDefault

**Reply by TkDodo** [reply]
I can take a closer look if you can reproduce it in a codesandbox.

**Comment by aubreyzulu** [comment]
Alright let me do that

**Comment by jaocfilho** [comment]
I have the same problem, I need to pass params to mutationFn. Did you manage to get it done?

**Reply by TkDodo** [reply]
> I can take a closer look if you can reproduce it in a codesandbox.

⬆️

**Comment by ajendrex** [comment]
Same here,

documentation [here](https://react-query-v3.tanstack.com/guides/mutations#persist-mutations) looks great, but I was not able to follow that pattern. The call to `mutate()` with parameters is invalid, the error says: "TS3245: Argument of type '\<whatever type my mutationFn receives\>' is not assignable to parameter of type 'void'"

**Reply by TkDodo** [reply]
if you use a defaultMutationFunction, there is no way that `useMutation` can actually infer the types of the mutationFunction. You'd likely have to help it out by providing the generics:

```
const client = new QueryClient();
client.setMutationDefaults(["addTodo"], {
  mutationFn: (newTodo: string) => axios.post("/api/data", { text: newTodo })
});

const addTodoMutation = useMutation<Todo, Error, string>({
  mutationKey: ["addTodo"],
  onSettled: () => {
    queryClient.invalidateQueries(["todos"]);
  }
});
```

- First generic is `TData` - the result type of the mutationFn.
- Second generic is `TError` - the error type (can be unknown)
- Third generic is `TVariables` - the type of the variables to pass in

so yeah you lose type inference but that is expected because setting something globally is something typescript cannot really know about.

**Reply by ajendrex** [reply]
thanks @TkDodo, that makes sense, I'll try it soon 👍
